This patch is not needed for GnuSASL 0.2.27.

--- gsasl-0.2.26/lib/digest-md5/server.c	2008-04-11 10:38:26.000000000 +0100
+++ local-gsasl-0.2.26/lib/digest-md5/server.c	2008-05-03 15:06:35.000000000 +0100
@@ -229,7 +229,7 @@
 	{
 	  /* Client provided username in ISO-8859-1 form, convert it
 	     to UTF-8 since the library is all-UTF-8. */
-	  char *username = latin1toutf8 (username);
+	  char *username = latin1toutf8 (state->response.username);
 	  if (!username)
 	    return GSASL_MALLOC_ERROR;
 	  gsasl_property_set (sctx, GSASL_AUTHID, username);
